cmd/snap-repair/runner.go

```diff
@@ -64,22 +76,139 @@ func (r *Repair) SetStatus(status RepairStatus) {
 	r.run.SaveState()
 }
 
+// nextCounter finds the next counter prefix to use for the output and status
+// files of a repair run
+func nextCounter(r *Repair) int {
+	counter := 0
+
+	// format: "r99.000001.2006-01-02T150405.output"
+	prefix := fmt.Sprintf("r%d.", r.Revision())
+	suffix := fmt.Sprintf(".output")
+	li, err := ioutil.ReadDir(r.RunDir())
+	if err != nil {
+		// should never happen
+		return counter
+	}
+	// ioutil.ReadDir() is already sorted
+	for _, dirent := range li {
+		name := dirent.Name()
+		if strings.HasPrefix(name, prefix) && strings.HasSuffix(name, suffix) {
+			li := strings.Split(name, ".")
+			if c, err := strconv.Atoi(li[1]); err == nil {
+				counter = c
+			}
+		}
+	}
+
+	return counter + 1
+}
+
 // Run executes the repair script leaving execution trail files on disk.
 func (r *Repair) Run() error {
-	// XXX initial skeleton...
 	// write the script to disk
-	rundir := filepath.Join(dirs.SnapRepairRunDir, r.BrandID(), r.RepairID())
+	rundir := r.RunDir()
 	err := os.MkdirAll(rundir, 0775)
 	if err != nil {
 		return err
 	}
+
 	script := filepath.Join(rundir, fmt.Sprintf("script.r%d", r.Revision()))
-	err = osutil.AtomicWriteFile(script, r.Body(), 0600, 0)
+	err = osutil.AtomicWriteFile(script, r.Body(), 0700, 0)
+	if err != nil {
+		return err
+	}
+
+	// the date may be broken so we use an additional counter
+	counter := nextCounter(r)
+	now := time.Now().Format("2006-01-02T150405")
+	baseName := fmt.Sprintf("r%d.%06d.%v", r.Revision(), counter, now)
+
+	logPath := filepath.Join(rundir, baseName+".output")
+	logf, err := os.OpenFile(logPath, os.O_RDWR|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+	defer logf.Close()
+
+	// run things and captures output etc
+	stR, stW, err := os.Pipe()
 	if err != nil {
 		return err
 	}
+	defer stR.Close()
+	defer stW.Close()
+
+	// run the script
+	env := os.Environ()
+	// we need to hardcode FD=3 because this is the FD after
+	// exec.Command() forked. there is no way in go currently
+	// to run something right after fork() in the child to
+	// know the fd. However because go will close all fds
+	// except the ones in "cmd.ExtraFiles" we are safe to set "3"
+	env = append(env, "SNAP_REPAIR_STATUS_FD=3")
+	env = append(env, "SNAP_REPAIR_RUN_DIR="+rundir)
+
+	workdir := filepath.Join(rundir, "work")
+	if err := os.MkdirAll(workdir, 0700); err != nil {
+		return err
+	}
+
+	cmd := exec.Command(script)
+	cmd.SysProcAttr = &syscall.SysProcAttr{Setpgid: true}
+	cmd.Env = env
+	cmd.Dir = workdir
+	cmd.ExtraFiles = []*os.File{stW}
+	cmd.Stdout = logf
+	cmd.Stderr = logf
+	if err = cmd.Start(); err != nil {
+		return err
+	}
+	stW.Close()
+
+	// wait for repair to finish or timeout
+	killTimerCh := time.After(defaultRepairTimeout)
+	doneCh := make(chan struct{})
+	go func() {
+		// ignore error, we only care about what we got via
+		// the status pipe
```

Comment:
thinking about this a bit more, do we really want to accept Done even if the exit code is != 0 ?

Reply:
It feels alright to ignore the status.. if the script isn't actually done it should really not be claiming to be.